Document ErrorGuaranteed (#1316)
rust-lang/rustc-dev-guide: src/diagnostics/error-guaranteed.md

# `ErrorGuaranteed`

The previous sections have been about the error message that a user of the
compiler sees. But emitting an error can also have a second important side
effect within the compiler source code: it generates an
[`ErrorGuaranteed`][errorguar].

`ErrorGuaranteed` is a zero-sized type that is unconstructable outside of the
[`rustc_errors`][rerrors] crate. It is generated whenever an error is reported
to the user, so that if your compiler code ever encounters a value of type
`ErrorGuaranteed`, the compilation is _statically guaranteed to fail_. This is
useful for avoiding unsoundness bugs because you can statically check that an
error code path leads to a failure.

There are some important considerations about the usage of `ErrorGuaranteed`:

* It does _not_ convey information about the _kind_ of error. For example, the

>   error may be due (indirectly) to an ICE. Thus, you should not rely on

Reviewer: I don't understand quite what this means! If there was an ICE, then how can we still be executing?

Author: I think I meant a delay span bug or other compiler error. Tbh, this is a bit cold in the cache for me...

Reviewer: Understandable :)

Reviewer: If you want to merge it, I'm ok with that, something is better than nothing!

Author: ```suggestion
  error may be due (indirectly) to a `delay_span_bug` or other compiler error.
  Thus, you should not rely on
```

>   `ErrorGuaranteed` in deciding whether to emit an error or what kind of error

Reviewer 2: ```suggestion
  `ErrorGuaranteed` when deciding whether to emit an error, or what kind of error
```

  to emit.

> * `ErrorGuaranteed` should not be used to indicated that a compilation _will

Reviewer 2: ```suggestion
* `ErrorGuaranteed` should not be used to indicate that a compilation _will
```

> * `ErrorGuaranteed` should not be used to indicated that a compilation _will
>   emit_ an error in the future. It should be used to indicate that an error

Reviewer 2: ```suggestion
* `ErrorGuaranteed` should not be used to indicated that a compilation will
  emit an error in the future. It should be used to indicate that an error
```

Author: Why?

Reviewer 2: Just a stylistic suggestion. :)
